File: 0042_trapping_rain_water/solution.cpp
class Solution {
public:
    int trap(vector<int>& towers) {
        vector<int> leftMax, rightMax;
        int num_towers = towers.size();
        if(num_towers == 0)
            return 0;
        int max_left = 0;
        int max_right = 0;
        for(int i = 0; i < num_towers; i++){
            max_left = max(towers[i], max_left);
            leftMax.push_back(max_left);
            max_right = max(towers[num_towers - i - 1], max_right);
            rightMax.push_back(max_right);
        }

        int water = 0;
        int curr_height;
        for(int i = 0; i < num_towers; i++){
            curr_height = towers[i];
            water += max(min(leftMax[i], rightMax[num_towers - i - 1]) - curr_height, 0);
        }
        return water;
    }
};

File: 0042_trapping_rain_water/solution.py
class Solution:
    def trap(self, towers: List[int]) -> int:
        num_towers = len(towers)
        maxLeft = list()
        maxRight = list()
        left_max = 0
        right_max = 0
        for index, tower_height in enumerate(towers):
            left_max = max(towers[index], left_max)
            maxLeft.append(left_max)
            right_max = max(towers[num_towers - index - 1], right_max)
            maxRight.append(right_max)

        water = 0
        for index, tower_height in enumerate(towers):
            water += max(min(maxLeft[index], maxRight[num_towers - index - 1]) - tower_height, 0)

        return water

